### PR TITLE
Move from kingpin to kong for Core and Rbac commands

### LINKED ISSUE #1953: Use a consistent CLI framework
### What problem are you facing?

When the new Crossplane CLI was introduced we chose to use [kong](https://github.com/alecthomas/kong), which is the successor to [kingpin](https://github.com/alecthomas/kingpin). However, core Crossplane and the RBAC managed still use kingpin (as well as all providers). 

### How could Crossplane help solve your problem?

Using a consistent CLI framework requires less contextual knowledge for maintainers and new contributors, and kingpin is now in maintenance mode. We should move Crossplane and the RBAC manager to using kong. We could also consider using kong in each of the providers, but that is less of a priority.

### PR DESCRIPTION
**Pull request by weastel**
Signed-off-by: Drumil Patel <drumilpatel720@gmail.com>

### Description of your changes

Fixes #1953 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Following features were tested:

1. Basic Structure of CLI
    You can refer following photos     
Screenshot from 2021-02-16 18-18-12
Screenshot from 2021-02-16 18-20-02
Screenshot from 2021-02-16 18-20-11

2. Debugging
   You can refer following photos
Screenshot from 2021-02-16 18-21-51
Screenshot from 2021-02-16 18-22-19
    
3. Enum Support  
    You can refer following photos
Screenshot from 2021-02-16 18-24-18

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: cmd/crossplane/rbac/rbac.go [hunk 1]
```diff
@@ -36,28 +35,19 @@ const (
...
+type Cmd struct {
+	Sync                time.Duration `short:"s" default:"1h" help:"Controller manager sync period duration such as 300ms, 1.5h or 2h45m."`
+	LeaderElection      bool          `short:"l" default:"false" env:"LEADER_ELECTION" help:"Use leader election for the conroller manager."`
+	ManagementPolicy    string        `short:"m" name:"manage" default:"All" enum:"All,Basic" help:"RBAC management policy."`
+	ProviderClusterRole string        `help:"A ClusterRole enumerating the permissions provider packages may request."`
 }
 
 // Run the RBAC manager.
-func (c *Command) Run(log logging.Logger) error {
+func (c *Cmd) Run(zl *logr.Logger) error {
```
Comment (alecthomas) [inline cmd/crossplane/rbac/rbac.go:49]: You should be able to use `BindTo()` to receive an interface here rather than a pointer.

**Comment by negz** [comment]
Thanks for tackling this @weastel! @hasheddan and @kasey would either of you be interested in reviewing this? I recall you dug into Kong to build crank.

**Comment by hasheddan** [comment]
@negz sure thing 🙂

**Comment by negz** [comment]
Hi @weastel! It seems like this change is failing on CI - can you try running `make reviewable` and making sure that any changes it generates are included in this PR? Can you also please fill out the "how has this code been tested" section of the PR template - I'd like to understand how you've verified that these changes work. We're going into feature freeze for Crossplane 1.1 next week, so we'll need to get this merged before Monday for it to make the cut.

**Review by alecthomas — APPROVED** [review]

**Comment by weastel** [comment]
@alecthomas I have some queries related to certain implementation. I tried using `kong.Vars` for adding enums but not sure why it is throwing an error. It would be helpful if you help me to resolve it, It will help me eliminate hard coding of enums. 
 Like [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/rbac/rbac.go#L44), I have defined Kong vars [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/main.go#L61).

**Comment by alecthomas** [comment]
> @alecthomas I have some queries related to certain implementation. I tried using `kong.Vars` for adding enums but not sure why it is throwing an error. It would be helpful if you help me to resolve it, It will help me eliminate hard coding of enums.
> Like [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/rbac/rbac.go#L44), I have defined Kong vars [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/main.go#L61).

I don't know what the error is, so I can't really say.

**Comment by weastel** [comment]
> > @alecthomas I have some queries related to certain implementation. I tried using `kong.Vars` for adding enums but not sure why it is throwing an error. It would be helpful if you help me to resolve it, It will help me eliminate hard coding of enums.
> > Like [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/rbac/rbac.go#L44), I have defined Kong vars [here](https://github.com/weastel/crossplane/blob/1659f2eb479ba2b8f7093241f4b8e263619354fb/cmd/crossplane/main.go#L61).
> 
> I don't know what the error is, so I can't really say.

Sorry my bad :disappointed:

**Comment by weastel** [comment]
> Hi @weastel! It seems like this change is failing on CI - can you try running `make reviewable` and making sure that any changes it generates are included in this PR? Can you also please fill out the "how has this code been tested" section of the PR template - I'd like to understand how you've verified that these changes work. We're going into feature freeze for Crossplane 1.1 next week, so we'll need to get this merged before Monday for it to make the cut.

@negz Sorry for the delay. I am unsure why e2e-tests are failing. Can somebody please help me to resolve this problem?

**Review by hasheddan — COMMENTED** [review]
@weastel apologies for the delay here, I ran integration tests on this branch locally successfully, so I have re-triggered the action to see if it was a flake. In the mean time, could you squash commits and rebase on `master`?

File: cmd/crossplane/rbac/rbac.go (hunk 1 above)
Comment (hasheddan) [reply cmd/crossplane/rbac/rbac.go:49]: @weastel Can you update here to receive an interface?

Comment (weastel) [reply cmd/crossplane/rbac/rbac.go]: @hasheddan @alecthomas I tried to receive an interface but got following error when running any command(For eg: `./crossplane core --debug`, `./crossplane core`):
```
crossplane: error: couldn't find binding of type logging.Logger for parameter 0 of core.Cmd.Run(), use kong.Bind(logging.Logger)
```

Is anything wrong in any implementations? If so can you please help me to resolve it.

**Comment by negz** [comment]
Hi @weastel - are you still interested in working on this? If so, could you please squash your commits and rebase on master per @hasheddan's comment and https://github.com/crossplane/crossplane/blob/master/CONTRIBUTING.md#contributing-code

**Review by hasheddan — COMMENTED** [review]
@weastel checking in again here :) Are you interested in continuing this work or would you like to pass off to someone else?

**Comment by weastel** [comment]
@hasheddan @negz Sorry I cannot complete this PR, I am busy with something else. You can close this PR by now or If possible you can pass it too. Apologies for not contributing to this PR.

**Comment by hasheddan** [comment]
@weastel no worries! We appreciate you letting us know and we will be happy to address this in another PR :) Best of luck with your other work!

**Pull request closed without merging**